torch/autograd/_functions/reduce.py

```diff
@@ -71,7 +71,10 @@ def safe_zeros_backward(inp, dim):
             exclusive_normal = exclusive_normal_nocp.cumprod(dim)
 
             def reverse_dim(var, dim):
-                return var.index_select(dim, Variable(torch.arange(var.size(dim) - 1, -1, -1)).long())
+                index = Variable(torch.arange(var.size(dim) - 1, -1, -1)).long()
+                if var.is_cuda:
+                    index = index.cuda()
```

Comment:
Exactly what I'm looking for, it seems more compact. 
But this recipe still doesn't work if the current device is different from the device you want to send the `index` to.

You can try this:
```python
var = torch.rand(5,5).cuda(1)
# ensure the current deviece
# torch.cuda.set_current_device(0)
torch.arange(var.size(dim) -1, -1, -1, out=var.data.new().long()) # Error raised
```